File: frontend/src/collab/protocol.ts
/**
 * 前端协议编解码与运行时校验。
 *
 * 服务端消息在进入任何业务逻辑之前必须经过这里：格式、协议版本与大小都在
 * 这一层拒绝，因此 provider 内部不会遇到半可信的对象。
 */

export const PROTOCOL_VERSION = 1

/** 单个 CRDT 更新的上限。 */
export const MAX_UPDATE_BYTES = 1024 * 1024
/** 单个文本帧的上限。 */
export const MAX_FRAME_BYTES = 2 * 1024 * 1024
/** 待发送更新累计字节的上限，超过后暂停新增编辑并保留已有记录。 */
export const MAX_PENDING_BYTES = 8 * 1024 * 1024
/** 服务端单连接发送队列的帧数上限。 */
export const MAX_OUTGOING_FRAMES = 256

export type Envelope = {
  v: typeof PROTOCOL_VERSION
  documentId: string
  syncId: string
}

export type ClientMessage = Envelope &
  (
    | { type: 'hello'; stateVector: string }
    | { type: 'tx'; txId: string; kind: 'edit' | 'catchup'; update: string }
    | { type: 'sync-end'; barrierId: string }
  )

/** 线上形状：二进制字段是 Base64 字符串。 */
export type ServerMessage = Envelope &
  (
    | { type: 'sync'; update: string; stateVector: string; seq: number }
    | { type: 'update'; update: string; seq: number }
    | { type: 'ack'; txId: string; seq: number }
    | { type: 'ready'; barrierId: string; seq: number }
    | { type: 'error'; code: string; retryable: boolean; message: string }
  )

/** 校验并解码后的形状：二进制字段已是字节，可直接使用。 */
export type DecodedServerMessage = Envelope &
  (
    | { type: 'sync'; update: Uint8Array; stateVector: Uint8Array; seq: number }
    | { type: 'update'; update: Uint8Array; seq: number }
    | { type: 'ack'; txId: string; seq: number }
    | { type: 'ready'; barrierId: string; seq: number }
    | { type: 'error'; code: string; retryable: boolean; message: string }
  )

export class ProtocolViolation extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ProtocolViolation'
  }
}

export function encodeHello(
  documentId: string,
  syncId: string,
  stateVector: Uint8Array,
): string {
  return JSON.stringify({
    v: PROTOCOL_VERSION,
    type: 'hello',
    documentId,
    syncId,
    stateVector: toBase64(stateVector),
  } satisfies ClientMessage)
}

export function encodeTx(
  documentId: string,
  syncId: string,
  txId: string,
  kind: 'edit' | 'catchup',
  update: Uint8Array,
): string {
  return JSON.stringify({
    v: PROTOCOL_VERSION,
    type: 'tx',
    documentId,
    syncId,
    txId,
    kind,
    update: toBase64(update),
  } satisfies ClientMessage)
}

export function encodeSyncEnd(
  documentId: string,
  syncId: string,
  barrierId: string,
): string {
  return JSON.stringify({
    v: PROTOCOL_VERSION,
    type: 'sync-end',
    documentId,
    syncId,
    barrierId,
  } satisfies ClientMessage)
}

export function toBase64(bytes: Uint8Array): string {
  let binary = ''
  // 分块拼接，避免一次展开超长参数列表。
  const chunk = 0x8000
  for (let index = 0; index < bytes.length; index += chunk) {
    binary += String.fromCharCode(...bytes.subarray(index, index + chunk))
  }
  return btoa(binary)
}

export function fromBase64(value: string): Uint8Array {
  let binary: string
  try {
    binary = atob(value)
  } catch {
    throw new ProtocolViolation('字段不是有效的 Base64')
  }
  const bytes = new Uint8Array(binary.length)
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index)
  }
  return bytes
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function requireString(source: Record<string, unknown>, key: string): string {
  const value = source[key]
  if (typeof value !== 'string') throw new ProtocolViolation(`${key} 必须是字符串`)
  return value
}

function requireSeq(source: Record<string, unknown>): number {
  const value = source.seq
  if (typeof value !== 'number' || !Number.isInteger(value) || value < 0) {
    throw new ProtocolViolation('seq 必须是非负整数')
  }
  return value
}

function requireUpdate(source: Record<string, unknown>, key: string): Uint8Array {
  const bytes = fromBase64(requireString(source, key))
  if (bytes.byteLength > MAX_UPDATE_BYTES) {
    throw new ProtocolViolation(`${key} 超过更新大小上限`)
  }
  return bytes
}

function requireType(source: Record<string, unknown>): string {
  const value = source.type
  if (typeof value !== 'string') throw new ProtocolViolation('缺少消息类型')
  return value
}

/**
 * 把收到的原始帧校验并解码为可直接使用的消息。
 *
 * 所有失败都抛 ProtocolViolation：调用方据此报错并关闭本连接，
 * 而不是把半解析的对象继续往下传。
 */
export function decodeServerMessage(data: unknown): DecodedServerMessage {
  if (typeof data !== 'string') throw new ProtocolViolation('只接受文本帧')
  if (data.length > MAX_FRAME_BYTES) throw new ProtocolViolation('帧超过大小上限')

  let parsed: unknown
  try {
    parsed = JSON.parse(data)
  } catch {
    throw new ProtocolViolation('帧不是有效 JSON')
  }
  if (!isRecord(parsed)) throw new ProtocolViolation('帧顶层必须是对象')

  if (parsed.v !== PROTOCOL_VERSION) {
    throw new ProtocolViolation(`不支持的协议版本 ${String(parsed.v)}`)
  }
  const documentId = requireString(parsed, 'documentId')
  const syncId = requireString(parsed, 'syncId')
  const type = requireType(parsed)
  const envelope = { v: PROTOCOL_VERSION, documentId, syncId } as const

  switch (type) {
    case 'sync':
      return {
        ...envelope,
        type,
        update: requireUpdate(parsed, 'update'),
        stateVector: requireUpdate(parsed, 'stateVector'),
        seq: requireSeq(parsed),
      }
    case 'update':
      return {
        ...envelope,
        type,
        update: requireUpdate(parsed, 'update'),
        seq: requireSeq(parsed),
      }
    case 'ack':
      return { ...envelope, type, txId: requireString(parsed, 'txId'), seq: requireSeq(parsed) }
    case 'ready':
      return {
        ...envelope,
        type,
        barrierId: requireString(parsed, 'barrierId'),
        seq: requireSeq(parsed),
      }
    case 'error': {
      const retryable = parsed.retryable
      if (typeof retryable !== 'boolean') {
        throw new ProtocolViolation('error.retryable 必须是布尔值')
      }
      return {
        ...envelope,
        type,
        code: requireString(parsed, 'code'),
        retryable,
        message: requireString(parsed, 'message'),
      }
    }
    default:
      throw new ProtocolViolation(`未知消息类型 ${type}`)
  }
}

File: backend/app/protocol.py
"""WebSocket 协议：消息编解码、大小限制与错误码。

线上格式固定为 JSON 信封加 Base64 二进制字段，便于调试；代价是额外体积。
本模块是唯一做格式校验的地方：房间与传输层只接收这里解析出的类型化对象，
因此无效消息不会带着半解析状态流入 CRDT 或数据库。

Pydantic 模型直接使用线上字段名（camelCase），避免别名映射带来的偏差。
"""

from __future__ import annotations

import base64
import binascii
import json
from dataclasses import dataclass
from enum import Enum
from typing import Annotated, Literal, Union
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StrictInt, TypeAdapter, ValidationError

PROTOCOL_VERSION = 1

#: 单个 CRDT 更新的上限。候选完整状态也受此约束，保证它仍能通过重连通道传输。
MAX_UPDATE_BYTES = 1024 * 1024
#: 单个 WebSocket 文本帧的上限。
MAX_FRAME_BYTES = 2 * 1024 * 1024
#: 客户端待发送更新的累计上限（前端使用）。
MAX_PENDING_BYTES = 8 * 1024 * 1024
#: 单个连接发送队列的帧数上限，超过则关闭该慢连接并由重连补同步。
MAX_OUTGOING_FRAMES = 256

#: 无法从帧中取得有效 syncId 时使用的占位符。
NIL_SYNC_ID = "00000000-0000-0000-0000-000000000000"


class ErrorCode(str, Enum):
    BAD_MESSAGE = "BAD_MESSAGE"
    UNSUPPORTED_VERSION = "UNSUPPORTED_VERSION"
    DOCUMENT_NOT_FOUND = "DOCUMENT_NOT_FOUND"
    TX_PAYLOAD_MISMATCH = "TX_PAYLOAD_MISMATCH"
    UPDATE_TOO_LARGE = "UPDATE_TOO_LARGE"
    INVALID_UPDATE = "INVALID_UPDATE"
    STORAGE_UNAVAILABLE = "STORAGE_UNAVAILABLE"
    ROOM_UNAVAILABLE = "ROOM_UNAVAILABLE"


#: 只有存储与房间故障可以自动重试；其余都是协议或数据错误，盲目重发没有意义。
RETRYABLE_CODES = frozenset({ErrorCode.STORAGE_UNAVAILABLE, ErrorCode.ROOM_UNAVAILABLE})


class ProtocolError(Exception):
    """携带固定错误码的协议异常。"""

    def __init__(self, code: ErrorCode, message: str) -> None:
        super().__init__(f"{code.value}: {message}")
        self.code = code
        self.message = message

    @property
    def retryable(self) -> bool:
        return self.code in RETRYABLE_CODES


# --- 线上消息模型 -----------------------------------------------------------


class _WireModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class _Envelope(_WireModel):
    v: StrictInt
    documentId: UUID
    syncId: UUID


class _HelloWire(_Envelope):
    type: Literal["hello"]
    stateVector: str


class _TxWire(_Envelope):
    type: Literal["tx"]
    txId: UUID
    kind: Literal["edit", "catchup"]
    update: str


class _SyncEndWire(_Envelope):
    type: Literal["sync-end"]
    barrierId: UUID


_ClientWire = Annotated[
    Union[_HelloWire, _TxWire, _SyncEndWire],
    Field(discriminator="type"),
]
_CLIENT_ADAPTER: TypeAdapter[_ClientWire] = TypeAdapter(_ClientWire)


# --- 解析结果 ---------------------------------------------------------------


@dataclass(frozen=True)
class Hello:
    document_id: str
    sync_id: str
    state_vector: bytes


@dataclass(frozen=True)
class Transaction:
    document_id: str
    sync_id: str
    tx_id: str
    kind: Literal["edit", "catchup"]
    update: bytes


@dataclass(frozen=True)
class SyncEnd:
    document_id: str
    sync_id: str
    barrier_id: str


ClientFrame = Union[Hello, Transaction, SyncEnd]


def _decode_base64(value: str, field: str) -> bytes:
    try:
        return base64.b64decode(value, validate=True)
    except (binascii.Error, ValueError) as error:
        raise ProtocolError(ErrorCode.BAD_MESSAGE, f"{field} 不是有效的 Base64") from error


def parse_client_message(raw: str) -> ClientFrame:
    """把原始文本帧解析为类型化消息，任何偏差都以固定错误码拒绝。"""
    if len(raw.encode("utf-8")) > MAX_FRAME_BYTES:
        raise ProtocolError(ErrorCode.BAD_MESSAGE, "帧超过大小上限")

    try:
        payload = json.loads(raw)
    except json.JSONDecodeError as error:
        raise ProtocolError(ErrorCode.BAD_MESSAGE, "帧不是有效 JSON") from error
    if not isinstance(payload, dict):
        raise ProtocolError(ErrorCode.BAD_MESSAGE, "帧顶层必须是对象")

    version = payload.get("v")
    if isinstance(version, int) and not isinstance(version, bool) and version != PROTOCOL_VERSION:
        raise ProtocolError(
            ErrorCode.UNSUPPORTED_VERSION, f"不支持的协议版本 {version}"
        )

    try:
        message = _CLIENT_ADAPTER.validate_python(payload)
    except ValidationError as error:
        raise ProtocolError(ErrorCode.BAD_MESSAGE, "帧字段不合法") from error

    document_id = str(message.documentId)
    sync_id = str(message.syncId)

    if isinstance(message, _HelloWire):
        return Hello(
            document_id=document_id,
            sync_id=sync_id,
            state_vector=_decode_base64(message.stateVector, "stateVector"),
        )
    if isinstance(message, _TxWire):
        update = _decode_base64(message.update, "update")
        if len(update) > MAX_UPDATE_BYTES:
            raise ProtocolError(ErrorCode.UPDATE_TOO_LARGE, "更新超过大小上限")
        return Transaction(
            document_id=document_id,
            sync_id=sync_id,
            tx_id=str(message.txId),
            kind=message.kind,
            update=update,
        )
    return SyncEnd(
        document_id=document_id, sync_id=sync_id, barrier_id=str(message.barrierId)
    )


def _dump(**fields: object) -> str:
    return json.dumps(fields, ensure_ascii=False, separators=(",", ":"))


def encode_sync(
    document_id: str, sync_id: str, update: bytes, state_vector: bytes, seq: int
) -> str:
    return _dump(
        v=PROTOCOL_VERSION,
        type="sync",
        documentId=document_id,
        syncId=sync_id,
        update=base64.b64encode(update).decode("ascii"),
        stateVector=base64.b64encode(state_vector).decode("ascii"),
        seq=seq,
    )


def encode_update(document_id: str, sync_id: str, update: bytes, seq: int) -> str:
    return _dump(
        v=PROTOCOL_VERSION,
        type="update",
        documentId=document_id,
        syncId=sync_id,
        update=base64.b64encode(update).decode("ascii"),
        seq=seq,
    )


def encode_ack(document_id: str, sync_id: str, tx_id: str, seq: int) -> str:
    return _dump(
        v=PROTOCOL_VERSION,
        type="ack",
        documentId=document_id,
        syncId=sync_id,
        txId=tx_id,
        seq=seq,
    )


def encode_ready(document_id: str, sync_id: str, barrier_id: str, seq: int) -> str:
    return _dump(
        v=PROTOCOL_VERSION,
        type="ready",
        documentId=document_id,
        syncId=sync_id,
        barrierId=barrier_id,
        seq=seq,
    )


def encode_error(
    document_id: str,
    sync_id: str,
    code: ErrorCode,
    message: str,
    retryable: bool | None = None,
) -> str:
    if retryable is None:
        retryable = code in RETRYABLE_CODES
    return _dump(
        v=PROTOCOL_VERSION,
        type="error",
        documentId=document_id,
        syncId=sync_id,
        code=code.value,
        retryable=retryable,
        message=message,
    )
